### PR TITLE
New weak reflection implementation

### PR DESCRIPTION
**Pull request by stuartwdouglas**
This implementation uses a reachability handler to only register the
classes if they are reachable.

### CODE REVIEW

**Comment by stuartwdouglas** [comment]
I tested with the kubernetes-client integration test. Compile time went from 200s to 76s, binary size went from 135mb to 67mb, and the tests still passed.

**Review by geoand — APPROVED** [review]
LGTM!

File: test-framework/jacoco/deployment/src/main/java/io/quarkus/jacoco/deployment/JacocoProcessor.java [hunk 1]
```diff
@@ -57,6 +57,7 @@ void transformerBuildItem(BuildProducer<BytecodeTransformerBuildItem> transforme
                 + config.dataFile;
         System.setProperty("jacoco-agent.destfile",
                 dataFile);
+        System.setProperty("jacoco-agent.jmx", "true");
```
Comment (gsmet) [inline test-framework/jacoco/deployment/src/main/java/io/quarkus/jacoco/deployment/JacocoProcessor.java:60]: This doesn't seem related? I just want to confirm you want this in.

Comment (geoand) [reply test-framework/jacoco/deployment/src/main/java/io/quarkus/jacoco/deployment/JacocoProcessor.java:60]: I actually added the same comment, but I think I never published it...

Comment (stuartwdouglas) [reply test-framework/jacoco/deployment/src/main/java/io/quarkus/jacoco/deployment/JacocoProcessor.java:60]: Oops, that was part of something else I was playing with

**Comment by gsmet** [comment]
@zakkak could you run this PR with GraalVM 20.3? I just want to be on the safe side.

**Comment by stuartwdouglas** [comment]
I tested with Graal 21.2, the kube client integration tests binary was 1mb smaller, but took 62s instead of 76 to compile.

**Comment by geoand** [comment]
Not bad for a minor version bump

**Comment by stuartwdouglas** [comment]
I tested locally with 20.3 as well, tests pass and binary size implies that  this is working as expected.

**Comment by Sanne** [comment]
> I tested with the kubernetes-client integration test. Compile time went from 200s to 76s, binary size went from 135mb to 67mb, and the tests still passed.

That sounds awesome.

May I suggest to add some logging? In my experience when I experimented with the Caffeine reachavbility handler it was crucial for me to keep being able to understand things :)

Also N.B. the same callback can be triggered multiple times so you need to guard against that. It's important because the reflective registration might trigger the need to add an additional analysis pass, leading to higher build times.

e.g. 

https://github.com/Sanne/quarkus/blob/ff72ff355ffddaab38ae1eb2b3cf7a218a410e69/extensions/caffeine/runtime/src/main/java/io/quarkus/caffeine/runtime/graal/CacheConstructorsAutofeature.java#L26-L48

**Comment by metacosm** [comment]
I've tried to build a native binary with a sample application I use:
- with this PR: `time mvn clean package -Pnative:  930.54s user 40.02s system 247% cpu 6:32.47 total`/ binary size: 70MB
- with 2.1.0.Final: `time mvn clean package -Pnative  1840.49s user 91.44s system 222% cpu 14:29.43 total`/ binary size: 135MB

Limited testing of the app didn't reveal any obvious issues… so this looks like a nice win to me!

**Comment by stuartwdouglas** [comment]
I have added the logging and the guard against multiple invocation.

**Review by gsmet — APPROVED** [review]
Nice, thanks!

**Comment by Sanne** [comment]
:+1: thanks Stuart! Amazing improvement, we should use this pattern more frequently.

**Comment by geoand** [comment]
I don't know if the pattern is more applicable to be honest... The reason why it applies to the Kubernetes Client case is that the model is gigantic and contains Jackson annotations all over it (this leading to reflection possible being needed)

**Comment by Sanne** [comment]
Well yes clearly the benefit in this case is much more visible, but in generaly any code X needing to reflectively access Y, should only register Y when X is reachable.
Same for resources: if the code which is needing a resource is dead, then the resource isn't needed.

And since removing access to Y might trigger more code to be not be reachable, there's sometimes cascading significant effects from small optimisations. Sometimes you have two small optimisations that individually don't do much, but when both are applied significant portions of libraries become dead.

**Comment by geoand** [comment]
Good point

**Comment by galderz** [comment]
> 👍 thanks Stuart! Amazing improvement, we should use this pattern more frequently.

Very impressive improvement. A very interesting pattern too to reduce reflection registrations.
